sirepo/package_data/static/js/sirepo.js

```diff
@@ -1556,14 +1544,157 @@ SIREPO.app.service('validationService', function(utilities) {
 
 });
 
+SIREPO.app.factory('srCache', function(appState, $rootScope) {
+
+    // Browser side caching implemented using indexedDB
+    // - Caches sim frame responses
+    // - Allows cache to be cleared by simId or (simId, modelName)
+    // - Keeps updateTime on item access and deletes expired records at startup
+
+    const self = {};
+    const STORE = 'db';
+    const FRAME = 'frame';
+    // 30 days until expired
+    const EXPIRY_TIME = 30 * 24 * 60 * 60 * 1000;
+    let db = null;
+
+    const deleteKeys = (keys) => {
+        if (! keys.length) {
+            return;
+        }
+        withObjectStore('readwrite', (o) => {
+            for (const k of keys) {
+                o.delete(k);
+            }
+        });
+    };
+
+    const initializeDatabase = () => {
+        if (! window.indexedDB || ! SIREPO.authState.uiWebSocket) {
+            return;
+        }
+        const r = window.indexedDB.open('srCache', 2);
+        r.onsuccess = (event) => {
+            db = event.target.result;
+            removeOldRecords();
+        };
+        r.onupgradeneeded = (event) => {
+            db = event.target.result;
```

Comment:
Yes, the onupgradeneeded handler shouldn't be setting ```db```. It turns out the onsuccess() gets called as well when onupgradedneeded completes sucessfully, so that is the only spot that needs to set the db. I've updated the code for this.

```If the onupgradeneeded event exits successfully, the onsuccess handler of the open database request will then be triggered.```